### PR TITLE
chore: Improve application logs adding message context

### PR DESCRIPTION
**Pull request by leoluz**
Signed-off-by: Leonardo Luz Almeida <leonardo_almeida@intuit.com>

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Review by jannfis — APPROVED** [review]
LGTM.

Just a very opinionated nit, feel free to merge though :)

File: server/application/application.go [hunk 1]
```diff
@@ -437,7 +440,7 @@ func (s *Server) Get(ctx context.Context, q *application.ApplicationQuery) (*app
 			})
 			return err
 		}); err != nil {
-			log.Warnf("Failed to force refresh application details: %v", err)
+			log.Warnf("failed to force refresh application details: %v", err)
```
[the remark this answers is not in the document]
Comment (leoluz) [reply server/application/application.go:443]: I actually agree with you. 
To be frank, this shouldn't be a matter of opinion. We should probably have this defined somewhere. 
For example:
https://github.com/golang/go/wiki/CodeReviewComments#error-strings

Will revert the change in this line. 👍🏻

**Comment by leoluz** [comment]
After analysing failing unit-tests I realized that this PR could potentially introduce an API breaking change by returning wrapped errors. It seems grpc library doesn't play nicely with wrapped errors:
https://github.com/grpc/grpc-go/issues/2934

Improved the code to avoid introducing API breaking change.
Can you pls take another look @jannfis ?

File: server/application/application.go [hunk 2]
```diff
@@ -238,53 +238,53 @@ func (s *Server) queryRepoServer(ctx context.Context, a *v1alpha1.Application, a
...
 			return status.Errorf(codes.InvalidArgument, "application references project %s which does not exist", a.Spec.Project)
 		}
-		return err
+		return fmt.Errorf("error getting application's project: %w", err)
 	}
 
 	helmRepos, err := s.db.ListHelmRepositories(ctx)
 	if err != nil {
-		return err
+		return fmt.Errorf("error listing help repositories: %w", err)
```
Comment (crenshaw-dev) [inline server/application/application.go:266]: s/help/helm

File: server/application/application.go [hunk 3]
```diff
@@ -238,53 +238,53 @@ func (s *Server) queryRepoServer(ctx context.Context, a *v1alpha1.Application, a
...
 	}
 
 	permittedHelmRepos, err := argo.GetPermittedRepos(proj, helmRepos)
 	if err != nil {
-		return err
+		return fmt.Errorf("error retrieving permitter repos: %w", err)
```
Comment (crenshaw-dev) [inline server/application/application.go:271]: s/permitter/permitted

File: server/application/application.go [hunk 4]
```diff
@@ -1041,17 +1055,17 @@ func (s *Server) PatchResource(ctx context.Context, q *application.ApplicationRe
 	if err != nil {
 		// don't expose real error for secrets since it might contain secret data
 		if res.Kind == kube.SecretKind && res.Group == "" {
-			return nil, fmt.Errorf("failed to patch Secret %s/%s", res.Namespace, res.Name)
+			return nil, fmt.Errorf("failed to patch Secret %s/%s: %w", res.Namespace, res.Name, err)
```
Comment (crenshaw-dev) [inline server/application/application.go:1058]: Revert due to comment above. Might want to log the real error instead.

File: server/application/application.go [hunk 5]
```diff
@@ -1571,10 +1586,11 @@ func (s *Server) Rollback(ctx context.Context, rollbackReq *application.Applicat
 		},
 	}
 	a, err = argo.SetAppOperation(appIf, *rollbackReq.Name, &op)
-	if err == nil {
-		s.logAppEvent(a, ctx, argo.EventReasonOperationStarted, fmt.Sprintf("initiated rollback to %d", rollbackReq.GetId()))
+	if err != nil {
+		return a, fmt.Errorf("error setting app operation: %w", err)
```
Comment (crenshaw-dev) [inline server/application/application.go:1590]: I think you can explicitly return nil, like in the refactor above. `SetAppOperation` won't return a non-nil value with an error.

File: util/grpc/errors.go [hunk 6]
```diff
@@ -31,6 +31,20 @@ func gitErrToGRPC(err error) error {
 	return err
 }
 
+// UnwrapGRPCStatus will attempt to cast the given error into a grpc Status
+// object unwrapping all existing inner errors. Will return nil if none of the
+// nested errors can be casted.
+func UnwrapGRPCStatus(err error) *status.Status {
```
Comment (crenshaw-dev) [inline util/grpc/errors.go:37]: Is this unit-testable?

File: util/grpc/errors.go [hunk 7]
```diff
@@ -31,6 +31,20 @@ func gitErrToGRPC(err error) error {
...
+		return se.GRPCStatus()
+	}
+	e := errors.Unwrap(err)
+	if e == nil {
+		return nil
+	}
+	return UnwrapGRPCStatus(e)
+}
+
 func kubeErrToGRPC(err error) error {
```
Comment (crenshaw-dev) [inline util/grpc/errors.go:48]: Ditto, does this need test coverage?

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
There's no way to overstate what a big deal this is for debugging. Thanks for doing this.

Comments include mostly nits and one important change.

File: util/grpc/errors.go (hunk 7 above)
Comment (leoluz) [reply util/grpc/errors.go:48]: added

File: util/grpc/errors.go (hunk 6 above)
Comment (leoluz) [reply util/grpc/errors.go:37]: yes :)

**Review by crenshaw-dev — APPROVED** [review]
lgtm!

**Comment by crenshaw-dev** [comment]
Cherry-picked onto 2.4.

**Pull request merged**